### PR TITLE
Unifying the data gathering process for data providers

### LINKED ISSUE #69: Unifying the data gathering process for data providers
Data Providers like:
- `FirstCommit` 
- `ProjectStarted` 
- `NumberOfCommits`
- `UsesSignedCommits`

Use commit history. It takes a toll on the performance when `fosstars-rating-core` try to hit the `GitHub API` again and again for each data provider to get the same data. 

To avoid this redundant work, merge the data gathering method for all the above-mentioned data providers.

### PR DESCRIPTION
**Pull request by sourabhsparkala**
- Updated github-api to version 1.109 in pom.xml
- Added a new Singleton GitHubDataFetcher, class to pull the GitHub data once and store the data
- Updated the remaining GitHub specific data providers
- Update Test classes to reflect this change
- Introducing Cache in GitHubDataFetcher
- Introducing new DataCache interface, implemented in GitHubDataCache

This fixes #69

### CODE REVIEW

**Comment by artem-smotrakov** [comment]
To make it work with `MultipleSecurityRatingsCalculator`, we need to make `GitHubDataPull` store info about multiple projects. But here is one problem. If `MultipleSecurityRatingsCalculator` processes many projects (for example, while it build a report for well-known open-source projects), then storing data about all processed projects may cause OOM. On the other hand, we don't need to store info about all projects. How about updating `GitHubDataPull` to cache info for a limited number of projects (maybe just one)? Here is what it may look like:

```java
// renamed GitHubDataPull
public GitHubDataFetcher {

   private static final int CACHE_CAPACITY = 3; // although 1 is actually enough for now

    private static final GitHubDataFetcher INSTANCE;

    private final GitHub github;

    // use FIFO or LRU cache with CACHE_CAPACITY
    private final Map<GitHubProject, List<GHCommit>> commitsCache = ... 

    // use FIFO or LRU cache with CACHE_CAPACITY
    private final Map<GitHubProject, List<GHCommit>> repositoryCache = ... 

    public static final GitHubDataFetcher instance(GitHub github) {
        // ... singleton
    }

    public List<GHCommit> commitsOf(GitHubProject project) { ... }

    public GHRepository repositoryOf(GitHubProject project) { ... }
}
```

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataPull.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+  /**
+   * An interface to the GitHub API.
+   */
+  private GitHub github;
+
+  /**
+   * {@linkplain GitHubProject} containing GitHub project information.
+   */
+  private GitHubProject project;
```
[the remark this answers is not in the document]
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataPull.java:43]: Yes. Good point.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataPull.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+   * Sort the commits collection in the reverse order by commit date. That is the first element is
+   * the latest commit and the last element is the first commit.
+   * 
+   * @return reverse sorted List of type {@link GHCommit}.
+   * @throws IOException occurred during REST call to GitHub API.
+   */
+  private List<GHCommit> reverseSort(List<GHCommit> commits) throws IOException {
+    return commits.stream()
+        .sorted(compareByCommitDate().reversed())
+        .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataPull.java:147]: I tried both. Unfortunately, the `repository.listCommits().toList()` returns an `unmodifiable` list. Which will not be sorted with these 2 techniques.

**Comment by sourabhsparkala** [comment]
@artem-smotrakov  Have updated the code. Please let me know what you think?

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataPull.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataPull.java]: This should work:

```java
new ArrayList<>(repository.listCommits().asList()).sort(Collections.reverseOrder(BY_COMMIT_DATE));
```

I don't have a proof that it works faster than streams. The code above copies a list and then sort it. It's pretty simple and straightforward. Streams are more complex by may be more elegant. So, feel free to use streams if you like. But please replace `compareByCommitDate().reversed()` with a `static final`:

```java
private static final Comparator<GHCommit> BY_COMMIT_DATE = (firstCommit, secondCommit) -> {
      try {
        return firstCommit.getCommitDate().compareTo(secondCommit.getCommitDate());
      } catch (IOException e) {
        throw new UncheckedIOException("Could not get date from the commit.", e);
      }
    };
```
It's not necessary to create a new comparator each time since it's stateless.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+   * @return List of {@link GHCommit}.
+   * @throws IOException occurred during REST call to GitHub API.
+   */
+  private List<GHCommit> commitsFor(GitHubProject project) throws IOException {
+    if (commitsCache.containsKey(project)) {
+      Optional<List<GHCommit>> commits = commitsCache.get(project);
+      if (commits.isPresent()) {
+        return commits.get();
+      }
+    }
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:68]: The `get()` method doesn't return null, does it? This could be a bit shorter and without nested if-blocks:

```java
Optional<List<GHCommit>> commits = commitsCache.get(project);
if (commits.isPresent()) {
  return commits.get();
}
```

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  public static GitHubDataFetcher instance(GitHub github) {
+    if (INSTANCE == null) {
+      INSTANCE = new GitHubDataFetcher(github);
+    }
+    return INSTANCE;
+  }
+  
+  /**
+   * Gets the list of commits for a GitHub project repository.
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:56]: I think it's worth to mention how the commits are sorted.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+  public GHRepository repositoryFor(GitHubProject project) throws IOException {
+    if (repositoryCache.containsKey(project)) {
+      Optional<GHRepository> repository = repositoryCache.get(project);
+      if (repository.isPresent()) {
+        return repository.get();
+      }
+    }
+
+    GHRepository repository = github.getRepository(project.path());
+    repositoryCache.put(project, repository);
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:91]: Please specify an expiration date. I am not sure if `GHRepository` will stay the same forever.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  /**
+   * Gets the first commit of the repository.
+   * 
+   * @param project of type {@link GitHubProject}, which holds the project information.
+   * @return {@link GHCommit} with the project information.
+   * @throws IOException occurred during REST call to GitHub API.
+   */
+  public GHCommit firstCommitFor(GitHubProject project) throws IOException {
+    List<GHCommit> commits = commitsFor(project);
+    return commits.get(commits.size() - 1);
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:104]: Strictly speaking, an `ArrayIndexOutOfBoundsException` can happen here.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  /**
+   * Get a list of commits which comes after a specific date.
+   * 
+   * @param date which is used as boundary.
+   * @param project of type {@link GitHubProject}, which holds the project information.
+   * @return sub-list of type {@link GHCommit}.
+   * @throws IOException occurred during REST call to GitHub API.
+   */
+  public List<GHCommit> filterByAfterDateFor(Date date, GitHubProject project) throws IOException {
+    return commitsFor(project).stream().filter(commit -> {
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:145]: I am not sure that streams take into account the fact that the commits are sorted by the commit date. I think it's better to rewrite it with a simple for-loop. That would speed it up a bit mosts cases (the existing data providers check commits for last several months).

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesSignedCommits.java [hunk 8]
```diff
@@ -32,51 +26,23 @@
...
-   * For parsing JSON.
+   * The date after which all the commits would be considered.
    */
-  private static final ObjectMapper MAPPER = new ObjectMapper();
-
-  /**
-   * 1 year in millis.
-   */
-  private static final long DELTA = 365 * 24 * 60 * 60 * 1000L;
+  private static final Date REFERENCE_DATE = Date.from(Instant.now().minus(Duration.ofDays(365)));
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesSignedCommits.java:31]: Let's rename it to `ONE_YEAR_AGO`.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesSignedCommits.java [hunk 9]
```diff
@@ -121,111 +87,41 @@ protected UsesSignedCommits doUpdate(GitHubProject project, ValueSet values) {
...
    */
-  private static class Page {
-
-    private final boolean hasNextPage;
-    private final JsonNode commitNodes;
-
-    private Page(boolean hasNextPage, JsonNode commitNodes) {
-      this.hasNextPage = hasNextPage;
-      this.commitNodes = commitNodes;
+  public boolean isCommitVerified(GHCommit commit) {
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesSignedCommits.java:120]: This predicate is stateless, therefore it should be a `static final` field.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetchAdapter.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.Date;
+import java.util.List;
+import org.kohsuke.github.GHCommit;
+import org.kohsuke.github.GHRepository;
+import org.kohsuke.github.GitHub;
+
+/**
+ * GitHub data pull adapter which wraps the exposed static functions from {@link GitHubDataFetcher}.
+ */
+public class GitHubDataFetchAdapter {
```
[the remark this answers is not in the document]
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetchAdapter.java:14]: Can you add the following method to the data providers that use `GitHubDataFetchAdapter`?

```java
GitHubDataFetchAdapter gitHubDataFetchAdapter() {
    return GitHubDataFetchAdapter.instance(github);
}
```

Can you then mock it in the tests?

[the remark this answers is not in the document]
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetchAdapter.java:14]: I think this should work.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java (hunk 3 above)
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:68]: `isPresent()` should check for that. And anyways `commitsCache.get(project);` returns an `Optional.empty()` if the content is not there.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java (hunk 6 above)
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:104]: I agree.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java (hunk 7 above)
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:145]: I agree.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java (hunk 3 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:68]: Okay, let's make the code shorter then.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/DataCache.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package com.sap.sgs.phosphor.fosstars.data;
+
+import java.util.Date;
+import java.util.Optional;
+
+/**
+ * This is an interface of a cache that stores different data types for a specific type of objects.
+ */
+public interface DataCache<K, V> {
```
[the remark this answers is not in the document]
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/DataCache.java:9]: Just to make it clear, it is not necessary to do it in this pull request.

[the remark this answers is not in the document]
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/DataCache.java:9]: Yes, I will create an issue for this.

[the remark this answers is not in the document]
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/DataCache.java:9]: Sure.

[the remark this answers is not in the document]
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/DataCache.java:9]: Referred it here https://github.com/SAP/fosstars-rating-core/issues/128

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java [hunk 12]
```diff
@@ -61,27 +47,21 @@ protected FirstCommit doUpdate(GitHubProject project, ValueSet values) throws IO
...
-    cache.put(project, value);
-
+    cache.put(project, value, tomorrow());
     return value;
   }
+
+  /**
+   * Get the Singleton instance of {@link GitHubDataFetcher}.
+   */
+  public GitHubDataFetcher gitHubDataFetcher() {
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java:64]: Please move this method to `AbstractGitHubDataProvider` to avoid duplicate code.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java [hunk 13]
```diff
@@ -61,27 +47,21 @@ protected FirstCommit doUpdate(GitHubProject project, ValueSet values) throws IO
...
-    }
-
-    if (lastPage == null) {
-      return UnknownValue.of(FIRST_COMMIT_DATE);
+    Value<Date> value = FIRST_COMMIT_DATE.unknown();
+    Optional<GHCommit> firstCommit = gitHubDataFetcher().firstCommitFor(project);
+    if (firstCommit.isPresent()) {
+      value =
+        FIRST_COMMIT_DATE
+              .value(firstCommit.get().getCommitDate());
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java:55]: That should fit in a single line.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+   * Get a list of commits which comes after a specific date.
+   * 
+   * @param date which is used as boundary.
+   * @param project of type {@link GitHubProject}, which holds the project information.
+   * @return sub-list of type {@link GHCommit}.
+   * @throws IOException occurred during REST call to GitHub API.
+   */
+  public List<GHCommit> commitsAfter(Date date, GitHubProject project)
+      throws IOException {
+    List<GHCommit> subList = new ArrayList<GHCommit>();
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:142]: Please don't specify `GHCommit` twice:

```java
List<GHCommit> subList = new ArrayList<>();
```

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+    for (GHCommit commit : commitsFor(project)) {
+      if (commit.getCommitDate().before(date)) {
+        break;
+      }
+      subList.add(commit);
+    }
+
+    return Collections.unmodifiableList(subList);
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java:151]: `unmodifiableList` doesn't seem to be necessary here since you create a new list `subList`.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/PackageManagement.java [hunk 16]
```diff
@@ -142,15 +142,15 @@ protected PackageManagement doUpdate(GitHubProject project, ValueSet values) thr
     if (something.isPresent()) {
       return something.get();
     }
-
+
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/PackageManagement.java:145]: Please remove trailing spaces.

File: src/test/java/com/sap/sgs/phosphor/fosstars/data/github/UsesSignedCommitTest.java [hunk 17]
```diff
@@ -13,95 +14,106 @@
...
+    ShortInfo info = mock(ShortInfo.class);
+    when(commit.getCommitShortInfo()).thenReturn(info);
 
-    HttpEntity entity = mock(HttpEntity.class);
-    when(response.getEntity()).thenReturn(entity);
+    GHVerification verification = mock(GHVerification.class);
+    when(info.getVerification()).thenReturn(verification);
+    when(verification.isVerified()).thenReturn(true);
 
-    try (InputStream content = getClass().getResourceAsStream("UsesSignedCommitTrue.json")) {
```
Comment (artem-smotrakov) [inline src/test/java/com/sap/sgs/phosphor/fosstars/data/github/UsesSignedCommitTest.java:42]: Is this file still necessary? Please also check other ones which used to be used by the test.

Comment (sourabhsparkala) [reply src/test/java/com/sap/sgs/phosphor/fosstars/data/github/UsesSignedCommitTest.java:42]: Yup deleted.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/PackageManagement.java (hunk 16 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/PackageManagement.java:145]: @sourabhsparkala

File: src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCache.java [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(entries);
+  }
+
+  /**
+   * Clear cache.
+   */
+  public void cacheClear() {
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCache.java:86]: Let's rename it to `clear()`.

File: src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCache.java [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * This is a cache for data of GitHub projects.
+ */
+public class GitHubDataCache<T> implements DataCache<GitHubProject, T> {
+
+  /**
+   * A map of cache entries.
+   */
+  private final LRUMap<GitHubProject, GitHubData<T>> entries;
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCache.java:17]: Please move this field below the static ones.

File: src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcher.java [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+   * @return List of {@link GHCommit}.
+   * @throws IOException occurred during REST call to GitHub API.
+   */
+  public List<GHCommit> commitsFor(GitHubProject project) throws IOException {
+    Optional<List<GHCommit>> cachedCommits = commitsCache.get(project);
+    if (cachedCommits.isPresent()) {
+      return cachedCommits.get();
+    }
+
+    List<GHCommit> commits = new ArrayList<GHCommit>(commitsFor(repositoryFor(project)));
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcher.java:88]: Please don't specify `GHCommit` twice.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java [hunk 21]
```diff
@@ -61,27 +47,12 @@ protected FirstCommit doUpdate(GitHubProject project, ValueSet values) throws IO
...
+    Value<Date> value = FIRST_COMMIT_DATE.unknown();
+    Optional<GHCommit> firstCommit = gitHubDataFetcher().firstCommitFor(project, github);
+    if (firstCommit.isPresent()) {
+      value = FIRST_COMMIT_DATE.value(firstCommit.get().getCommitDate());
     }
-
-    Value<Date> value = FIRST_COMMIT_DATE.value(lastPage.get(lastPage.size() - 1).getCommitDate());
-    cache.put(project, value);
-
+    cache.put(project, value, tomorrow());
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java:55]: I see that some providers specify an expiration date and some rely on the cache. We'll need to think if it should be unified. We have a couple of issues to improve caching in the providers - let's work on it later in #128 and #100

File: src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcher.java [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+/**
+ * Helper class for GitHub data providers which pulls the data from a GitHub repository. Also, the
+ * class caches the fetched data for a certain amount of time.
+ */
+public class GitHubDataFetcher {
+
+  /**
+   * The singleton instance.
+   */
+  private static GitHubDataFetcher INSTANCE;
```
Comment (artem-smotrakov) [inline src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcher.java:24]: Since the instance doesn't need `GitHub` in the constructor, let's make the field final.

```java
private static final GitHubDataFetcher INSTANCE = new GitHubDataFetcher();
```

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java (hunk 21 above)
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/FirstCommit.java:55]: ok

File: src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+
+  @Test
+  public void testPut() throws IOException {
+    GitHubDataCache<String> gitHubDataCache = new GitHubDataCache<String>();
+    
+    String test = "test1";
+    GitHubProject project = new GitHubProject(new GitHubOrganization(test), test);
+    gitHubDataCache.put(project, test);
+    assertTrue(gitHubDataCache.size() == 1);
+    assertTrue(gitHubDataCache.get(project).get().equals(test));
```
Comment (artem-smotrakov) [inline src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java:18]: Please use `assertEquals()`.

File: src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    assertTrue(gitHubDataCache.size() == 1);
+    assertTrue(gitHubDataCache.get(project2).get().equals(test));
+
+    test = "test3";
+    GitHubProject project3 = new GitHubProject(new GitHubOrganization(test), test);
+    gitHubDataCache.put(project3, test);
+    assertTrue(gitHubDataCache.size() == 2);
+    assertTrue(gitHubDataCache.get(project3).get().equals(test));
+
+    test = "test4";
```
Comment (artem-smotrakov) [inline src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java:32]: Okay, here we know that `put()` works. The rest of checks looks like dups.

File: src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcherTest.java [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+    assertNotNull(actualRepo2);
+    assertTrue(gitHubDataFetcher.repositoryCache.size() == 1);
+    assertEquals(expectedRepo, actualRepo2);
+
+    GitHubProject project3 = mock(GitHubProject.class);
+    GHRepository actualRepo3 = gitHubDataFetcher.repositoryFor(project3, github);
+    assertNotNull(actualRepo3);
+    assertTrue(gitHubDataFetcher.repositoryCache.size() == 2);
+
+    GitHubProject project4 = mock(GitHubProject.class);
```
Comment (artem-smotrakov) [inline src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcherTest.java:43]: What do the rest of the checks check?

Comment (sourabhsparkala) [reply src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcherTest.java:43]: it checks the max capacity as well.

File: src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java (hunk 24 above)
Comment (sourabhsparkala) [reply src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java:32]: The remaining line checks for max capacity

File: src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcher.java (hunk 22 above)
Comment (sourabhsparkala) [reply src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataFetcher.java:24]: Agree 👍

File: src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java (hunk 23 above)
Comment (sourabhsparkala) [reply src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java:18]: Done

File: src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java (hunk 24 above)
Comment (artem-smotrakov) [reply src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/GitHubDataCacheTest.java:32]: Then, I think it should use a loop.
